### ISSUE TITLE
Generator with await not considered an AsyncGenerator

### ISSUE THREAD

**Issue by audoh-tickitto** [microsoft/pylance-release#1348]
## Environment data

-   Language Server version: 2021.5.3
-   OS and version: Ubuntu 20.04.2
-   Python version (& distribution if applicable, e.g. Anaconda): 3.9.2, acquired via pyenv

## Expected behaviour

- Inferred type should be AsyncGenerator
- Await should be allowed within a non-async function when within a generator

## Actual behaviour

- Inferred type is Generator.
- 'await' only allowed within async function

## Logs

N/A (should be reproducible from code snippet)

## Code Snippet / Additional information

```python
import asyncio
from typing import AsyncGenerator

async def get_value(v: int) -> int:
    await asyncio.sleep(1)
    return v + 1

def get_generator() -> AsyncGenerator[int, None]:
    return (await get_value(v) for v in [1, 2, 3])

async def test() -> None:
    print(type(get_generator()))
    async for x in get_generator():
        print(x)

asyncio.get_event_loop().run_until_complete(test())
```
**Python result:**
```
<class 'async_generator'>
2
3
4
```
**Pylance error:**
![image](https://user-images.githubusercontent.com/71696787/119673437-40850100-be33-11eb-8158-2456fb321b55.png)

**Note:** the exact same issues seem to be present in mypy:
https://github.com/python/mypy/issues/10534

**Comment by erictraut** [comment]
I'm surprised that this code runs without generating a runtime exception.

[PEP 492](https://www.python.org/dev/peps/pep-0492/#await-expression), which introduced the `async` and `await` keywords to the Python language explicitly states:

>It is a SyntaxError to use await outside of an async def function (like it is a SyntaxError to use yield outside of def function).

Likewise, [PEP 530](https://www.python.org/dev/peps/pep-0530/#await-in-comprehensions), which introduced the use of `await` in comprehensions, says:

>We propose to allow the use of await expressions in both asynchronous and synchronous comprehensions. This is only valid in `async def` function body.

Perhaps you can help me understand why the Python interpreter appears to allow the use of `await` in this example even though it's used outside of an `async def` function body?

**Comment by audoh-tickitto** [comment]
That aspect surprised me as well to be honest and is more likely just lenience in the Python interpreter allowing it to run technically invalid code as if it were valid. Originally the sample code did use an async function, but I removed it once I realised it still held up.

However, with regards to comprehensions - I believe that would be referring specifically to _list_ comprehensions, rather than generators - I don't believe there's any way a synchronous generator could involve an await, regardless of the function which called it.

Therefore, the issue of asynchronous generators still holds if you make the function async and ignore the weirdness of having an await in a non-async function:
```python
import asyncio
from typing import AsyncGenerator


async def get_value(v: int) -> int:
    await asyncio.sleep(1)
    return v + 1


async def get_generator() -> AsyncGenerator[int, None]:
    return (await get_value(v) for v in [1, 2, 3])


async def test() -> None:
    print(type(await get_generator()))
    async for x in await get_generator():
        print(x)


asyncio.get_event_loop().run_until_complete(test())
```

I'll revise the original issue accordingly.

**Comment by brettcannon** [comment]
I think the code was _technically_ valid for the `def spam(): return (await x for x in range(3))` because the generator expression isn't executed upon creation, so your synchronous function is simply returning an async generator that isn't run (yet), so to Python you're just returning an object which was created in isolation and will be call elsewhere. And thanks to async generators not being a plain generator, you can't execute it without using it in something like `async for` which will then trigger a check in Python for using an awaitable outside of an `async` function and cause the `TypeError`.

**Comment by audoh-tickitto** [comment]
Yes, actually I think that PEP, when it refers to comprehension, means specifically list/dict/set comprehension, and not generators.

**Comment by erictraut** [comment]
I've updated the type checking logic to allow these cases without generating false positive errors.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.6.0, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202160-2-june-2021
